### PR TITLE
[6/n] allowances: integration tests

### PR DESCRIPTION
**Pull request by manolisliolios**
## Description 

Tests a big list of variations for scenarios in transactional and e2e tests.

Combination of funder address balance + allowance spends, multi-allowance spends, edge cases, soft-bundle post-sign-time revocations, soft-bundle mid-transaction running out of gas, sponsor combos, etc.

## Test plan 

A huge set of tests for different happy/edge case paths

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] Indexing Framework:

### CODE REVIEW

File: crates/sui-transactional-test-runner/src/args.rs [hunk 1]
```diff
@@ -552,6 +560,62 @@ impl SuiExtraValueArgs {
...
+        ))
+    }
+
+    fn skip_whitespace<'a, I: Iterator<Item = (ValueToken, &'a str)>>(
+        parser: &mut MoveCLParser<'a, ValueToken, I>,
+    ) {
+        while let Some(ValueToken::Whitespace) = parser.peek_tok() {
+            let _ = parser.advance(ValueToken::Whitespace);
+        }
+    }
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/args.rs]: There shouldn't actually be whitespace here due to how clap separates arguments. Why was this added?

File: crates/sui-transactional-test-runner/src/args.rs [hunk 2]
```diff
@@ -603,15 +667,7 @@ impl SuiExtraValueArgs {
         // Parse the type from the type string
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/args.rs]: I'm surprised any of this exists

Comment (tnowacki) [reply crates/sui-transactional-test-runner/src/args.rs]: I shall pay my penance for accepting this monstrosity

Comment (tnowacki) [reply crates/sui-transactional-test-runner/src/args.rs]: Penance
https://github.com/MystenLabs/sui/pull/27582

File: crates/sui-transactional-test-runner/src/args.rs (hunk 1 above)
Comment (manolisliolios) [reply crates/sui-transactional-test-runner/src/args.rs]: Wasn't needed indeed, removed!

File: crates/sui-adapter-transactional-tests/tests/allowances/custom_coin.snap [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+unchanged_shared: 0x0000000000000000000000000000000000000000000000000000000000000006, 0x0000000000000000000000000000000000000000000000000000000000000403
+accumulators_written: (object(2,0), A, sui::balance::Balance<test::foo::FOO>, Split, 400), (object(5,0), B, sui::balance::Balance<test::foo::FOO>, Merge, 400)
+gas summary: computation_cost: 1000000, storage_cost: 4126800,  storage_rebate: 3107412, non_refundable_storage_fee: 31388
+
+task 6, line 43:
+//# create-checkpoint
+Checkpoint created: 2
+
+task 7, line 45:
+//# view-funds sui::balance::Balance<test::foo::FOO> A
```
Comment (tzakian) [inline crates/sui-adapter-transactional-tests/tests/allowances/custom_coin.snap:65]: Nit: maybe view funds before the checkpoint to show that this is because of the send funds above.

File: crates/sui-transactional-test-runner/src/args.rs [hunk 4]
```diff
@@ -570,6 +578,46 @@ impl SuiExtraValueArgs {
...
+                )
+            })?;
+
+        parser.advance(ValueToken::LParen)?;
+        let amount_str = parser.advance(ValueToken::Number)?;
+        let (amount, _) = parse_u64(amount_str)?;
+        parser.advance(ValueToken::Comma)?;
+        if let Some(ValueToken::AtSign) = parser.peek_tok() {
+            parser.advance(ValueToken::AtSign)?;
+        }
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/args.rs]: Why are we conditionally parsing `@`? I don't understand the intent here

File: crates/sui-adapter-transactional-tests/tests/allowances/app_allowance.snap [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+//# programmable --sender C --inputs allowance_withdraw<sui::balance::Balance<sui::sui::SUI>>(300,@A,object(4,0)) mutshared(4,0) immshared(6) @C
+// The new spender works.
+//> 0: test::app::app_pay<sui::sui::SUI>(Input(1), Input(0), Input(2), Input(3));
+mutated: object(0,2), object(4,0)
+unchanged_shared: 0x0000000000000000000000000000000000000000000000000000000000000006
+accumulators_written: (object(2,0), A, sui::balance::Balance<sui::sui::SUI>, Split, 300), (object(11,0), C, sui::balance::Balance<sui::sui::SUI>, Merge, 300)
+gas summary: computation_cost: 1000000, storage_cost: 4696800,  storage_rebate: 3671712, non_refundable_storage_fee: 37088
+
+task 12, lines 96-97:
+//# view-object 4,0
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/allowances/app_allowance.snap:101]: not sure how useful this is here but its fine

File: crates/sui-adapter-transactional-tests/tests/allowances/calendar_window.move [hunk 6]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// A monthly rate limit driven by the real Clock: the first spend anchors the
+// window, exhausting it blocks further spends, and advancing the clock past
+// the anniversary rolls the window over.
+
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/allowances/calendar_window.move:7]: I feel like we are missing a log around possible time range/windows confiturations

**Review by tnowacki — COMMENTED** [review]
Trying to read through coverage, do we have timestamp coverage? For start time and expiration?

File: crates/sui-transactional-test-runner/src/args.rs (hunk 4 above)
Comment (manolisliolios) [reply crates/sui-transactional-test-runner/src/args.rs]: Good q: it seems like when we use addresses, we do prefix them with `@` so made this non-optional. 

Was confused with the `--sender A` kind of flows where A does not have the prefix.

File: crates/sui-adapter-transactional-tests/tests/allowances/calendar_window.move (hunk 6 above)
Comment (manolisliolios) [reply crates/sui-adapter-transactional-tests/tests/allowances/calendar_window.move:7]: I intentionally skipped the move logic to avoid duplicating tests that are already there in 1/n.

Do you think it's worth adding transactional tests too for those?

**Review by tzakian — APPROVED** [review]
🧪
